test/SerialIOTest.cpp

```diff
@@ -83,6 +83,54 @@ TEST_CASE( "multi_series_test", "[serial]" )
     allSeries.clear();
 }
 
+TEST_CASE( "multiple_series_handles_test", "[serial]" )
+{
+    /*
+     * First test: No premature flushes through destructor when another copy
+     * is still around
+     */
+    {
+        std::unique_ptr< openPMD::Series > series_ptr;
+        {
+            openPMD::Series series(
+                "sample%T.json", openPMD::AccessType::CREATE );
+            series_ptr = std::unique_ptr< openPMD::Series >(
+                new openPMD::Series( series ) );
+            /*
+             * we have two handles for the same Series instance now:
+             * series and series_ptr
+             * series goes out of scope before series_ptr
+             * destructor should only do a flush after both of them go out of
+             * scope, but it will currently run directly after this comment
+             * since no iteration has been written yet, an error will be thrown
```

Comment:
perfect description, thanks!
looks like we should add reference counting...

Reply:
I'm still liking the idea of having a clear separation of:
1. Non-copyable shared internal resources
2. Copyable handles that are exposed to user code and that have shared pointers to the internal resources.

Currently, all of our objects just hide any member variable that need to be shared between instances behind shared pointers, which results in bugs like this.
Also, having *actual* internal classes for non-copyable resources would allow us to have proper upwards pointers through the openPMD hierarchies. Because, currently `Writable::parent` just goes to a copyable handle, and that concerns not only `Series` but *all* of our objects inheriting from `Attributable`. The issue is fortunately hard enough to trigger, but I should maybe try to find a way to make it fail for classes other than `Series` too...

> Not sure how useful an alias for a Series in form of a handle object

All other classes that represent entities in the openPMD hierarchies are designed as handles, too, (sharing similar issues as argued above), so it's a question of consistency imo

Reply:
>  or forbid the Series copy constructor...

Until we get to it, this would likely be a sane thing to do. (This is what I temporarily did when writing the openPMD plugin for PIConGPU in order to avoid this problem). Breaking change tho

Reply:
I agree with your rationale above.

Yes, deleting the copy constructor would just be a quick-fix, because it is in fact broken. (Move hopefully works? Rule of five would be good for this class #790).